src/agent/src/image_rpc.rs

```diff
@@ -33,11 +58,227 @@ pub struct ImageService {
     sandbox: Arc<Mutex<Sandbox>>,
 }
 
+#[derive(Serialize, Debug, Default, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct IndexDescriptor {
+    pub schema_version: u8,
+    pub manifests: Vec<OciDescriptor>,
+}
+
 impl ImageService {
     pub fn new(sandbox: Arc<Mutex<Sandbox>>) -> Self {
         Self { sandbox }
     }
 
+    fn build_oci_path(cid: &str) -> PathBuf {
+        let mut oci_path = PathBuf::from("/tmp");
+        oci_path.push(cid);
+        oci_path.push(IMAGE_OCI);
+        oci_path
+    }
+
+    fn handle_layer_data(
+        image_data: &ImageData,
+        image_manifest: &mut OciManifest,
+        crypto_config: &CryptoConfig,
+        oci_blob_path: &Path,
+    ) -> Result<()> {
+        for layer in image_data.layers.iter() {
+            let layer_digest = layer.clone().sha256_digest();
+
+            if layer.media_type == MEDIA_TYPE_LAYER_GZIP_ENC
+                || layer.media_type == MEDIA_TYPE_LAYER_ENC
+            {
+                if let Some(decrypt_config) = &crypto_config.decrypt_config {
+                    for layer_desc in image_manifest.layers.iter_mut() {
+                        if layer_desc.digest == layer_digest {
+                            let (layer_decryptor, _dec_digest) = decrypt_layer(
+                                decrypt_config,
+                                layer.data.as_slice(),
+                                layer_desc,
+                                false,
+                            )?;
+                            let mut plaintxt_data: Vec<u8> = Vec::new();
+                            let mut decryptor = layer_decryptor.unwrap();
+
+                            decryptor.read_to_end(&mut plaintxt_data)?;
+                            let layer_name = format!("{:x}", Sha256::digest(&plaintxt_data));
+                            let mut out_file = File::create(oci_blob_path.join(&layer_name))?;
+                            info!(sl!(), "Saving layer file {}...", layer_name);
+                            out_file.write_all(&plaintxt_data)?;
+                            layer_desc.media_type =
+                                manifest::IMAGE_LAYER_GZIP_MEDIA_TYPE.to_string();
+
+                            layer_desc.digest = format!("{}:{}", DIGEST_SHA256, layer_name);
+                        }
+                    }
+                }
+            } else if let Some(layer_name) =
+                layer_digest.strip_prefix(format!("{}:", DIGEST_SHA256).as_str())
+            {
+                let mut out_file = File::create(oci_blob_path.join(&layer_name))?;
+                info!(sl!(), "Saving layer file {}...", layer_name);
+                out_file.write_all(&layer.data)?;
+            } else {
+                error!(
+                    sl!(),
+                    "layer digest algo not supported:: {:?}", layer_digest
+                );
```

Comment:
I haven't checked, but this formatting looks odd: it's too spaced out. Are you using `rustfmt`?

Reply:
Yes, cargo fmt generated this format